src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,147 @@
 package cromwell.binding
 
+import cromwell.binding.AstTools.AstNodeName
+import cromwell.parser.WdlParser.{Ast, AstList, AstNode}
+
+import scala.collection.JavaConverters._
+
+
+object Scope {
+
+  /**
+   * Utility class to generate a ScopeTree from an Ast.
+   */
+  private class ScopeTreeGenerator {
+
+    /* Keep track of the # of created Scatter nodes. */
+    var scatterCount: Int = 0
+
+    /**
+     * Retrieve the list of AstNodes from the AST body.
+     * Return empty seq if body is null or is not a list.
+     */
+    private def getChildrenList(ast: Ast): Seq[AstNode] = Option(ast.getAttribute("body")) collect { case list: AstList => list.asScala.toVector } getOrElse Seq.empty
+
+    /**
+     * Generate recursively all descendant Scopes for the workflow.
+     * Should only be called during apply with a root Ast
+     * @return Some(instance of Scope) if the node has a Scope equivalent, None otherwise
+     */
+    def generateScopeTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Option[Scope] = {
+      node match {
+        case w: Ast if w.getName == AstNodeName.Workflow =>
+          val workflowChildren = getChildrenList(w) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter) }
+          val workflow: Workflow = Workflow(w, wdlSyntaxErrorFormatter)
+          workflow.setChildren(workflowChildren)
+          Some(workflow)
+
+        case c: Ast if c.getName == AstNodeName.Call => Some(Call(c, namespaces, tasks, wdlSyntaxErrorFormatter))
+
+        case s: Ast if s.getName == AstNodeName.Scatter =>
+          // Need to create the Scatter before generating its children to have the correct index layout order
+          val scatter = Scatter(s, scatterCount)
```

Comment:
Actually looking at it a bit more I'm thinking that the one to remove is the var parent. You're constructing the object before the recursive descent so you can pass that into the next iteration and any children at that level can be constructed w/ that as the parent